File: Java/strings/LongestCommonPrefix.java
//  This is Leetcode based DSA question --
// ✅ Problem Statement

/**
 * Longest Common Prefix
 *
 * Time Complexity: O(n * m)   // n = number of strings, m = length of shortest string
 * Space Complexity: O(1)
 *
 * Finds the longest prefix string common to all given strings.
 */

public class LongestCommonPrefix {

    /**
     * Finds the longest common prefix among a list of strings.
     *
     * @param strs Array of strings
     * @return     Longest common prefix
     */
    public String longestCommonPrefix(String[] strs) {
        if (strs == null || strs.length == 0) return "";

        String prefix = strs[0];

        for (int i = 1; i < strs.length; i++) {
            while (strs[i].indexOf(prefix) != 0) {
                prefix = prefix.substring(0, prefix.length() - 1);
                if (prefix.isEmpty()) return "";
            }
        }

        return prefix;
    }

    // Example usage
    public static void main(String[] args) {
        LongestCommonPrefix solver = new LongestCommonPrefix();
        String[] words = {"flower", "flow", "flight"};

        System.out.println("Longest Common Prefix: " + solver.longestCommonPrefix(words));
    }
}

File: C++/strings/longestcommon_prefix.cpp
/*
Problem: Longest Common Prefix
---------------------------------------------
Given an array of strings, find the longest common prefix 
among all strings. If there is no common prefix, return an empty string.

Example:
Input: ["flower", "flow", "flight"]
Output: "fl"

Input: ["dog", "racecar", "car"]
Output: ""  // No common prefix

Approach:
1️⃣ Sort the array of strings.
2️⃣ Compare the first and last strings only — since sorting arranges them lexicographically,
    their common prefix will be the common prefix of the entire array.

Time Complexity: O(n * m)
  where n = number of strings, m = length of the shortest string
Space Complexity: O(1)
*/

#include <bits/stdc++.h>
using namespace std;

string longestCommonPrefix(vector<string>& strs) {
    if (strs.empty()) return "";

    sort(strs.begin(), strs.end());
    string first = strs[0];
    string last = strs[strs.size() - 1];
    int i = 0;
    
    while (i < first.size() && i < last.size() && first[i] == last[i])
        i++;
        
    return first.substr(0, i);
}

int main() {
    vector<string> strs1 = {"flower", "flow", "flight"};
    vector<string> strs2 = {"dog", "racecar", "car"};

    cout << longestCommonPrefix(strs1) << endl;  // Output: fl
    cout << longestCommonPrefix(strs2) << endl;  // Output: (empty)

    return 0;
}
